core/src/main/scala/nl/vroste/zio/kinesis/client/zionative/Consumer.scala

```diff
@@ -426,6 +426,95 @@ object Consumer {
            }.runDrain
     } yield ()
 
+  /**
+   * Apply an effectful function to each record in a stream.
+   *
+   * This function ensures that a checkpoint is performed between every 2 records with the same partition key. This is
+   * useful when you want to ensure that records with the same partition key are processed in order. Even in the
```

Comment:
Can you explain how this guarantees ordering more than sequentially processing messages in a shard?

Reply:
If you process all records sequentially you will have full ordering guarantees. This improves on that in 2 cases:

1. You process records in a shard concurrently. Even in this case this checkpointer will guarantee that records with the same partition key are processed sequentially.
2. You persist state externally and kinesis fails. Imagine we are consuming 3 records and fail before being able to checkpoint. Meanwhile we are persiting state in some external db, let's say dynamo.

The first time you are processing the calles would look like this:
(event1, state1) -> (event2, state2) -> (event3, state3)
The second time you are processing (after the failure) the calls would look like this:
(event1, state3) -> (event2, state4) -> (event3, state5)

So you kind have to support replaying of an arbitrary prefix in the partition. There are some applications that care about this and where dealing with this behavior is annoying.

This checkpointer behaves differently and in the second case you would only see
(event3, state3).
It simplifies reasoning as you only ever need to deal with replays of the last event.

I'll adjust the docstring to explain this a bit better 👍

Reply:
Makes sense, I missed the parallelism part. Thanks for the detailed explanation!